pandas/core/sparse/array.py

```diff
@@ -446,161 +623,341 @@ def _get_val_at(self, loc):
+            blocs = np.concatenate(blocs)
+            blengths = np.concatenate(blengths)
+
+            sp_index = BlockIndex(length, blocs, blengths)
+
+        return cls(data, sparse_index=sp_index, fill_value=fill_value)
+
+    def astype(self, dtype=None, copy=True):
         """
-        Compute sum of non-NA/null observations in SparseArray. If the
-        fill_value is not NaN, the "sparse" locations will be included in the
-        observation count.
+        Change the dtype of a SparseArray.
+
+        The output will always be a SparseArray. To convert to a dense
+        ndarray with a certain dtype, use :meth:`numpy.asarray`.
+
+        Parameters
+        ----------
+        dtype : np.dtype or ExtensionDtype
+            For SparseDtype, this changes the dtype of
+            ``self.sp_values`` and the ``self.fill_value``.
+
+            For other dtypes, this only changes the dtype of
+            ``self.sp_values``.
+
+        copy : bool, default True
+            Whether to ensure a copy is made, even if not necessary.
 
         Returns
         -------
-        nobs : int
+        SparseArray
+
+        Examples
+        --------
+        >>> arr = SparseArray([0, 0, 1, 2])
+        >>> arr
+        [0, 0, 1, 2]
+        Fill: 0
+        IntIndex
+        Indices: array([2, 3], dtype=int32)
+
+        >>> arr.astype(np.dtype('int32'))
+        [0, 0, 1, 2]
+        Fill: 0
+        IntIndex
+        Indices: array([2, 3], dtype=int32)
+
+        Using a NumPy dtype with a different kind (e.g. float) will coerce
+        just ``self.sp_values``.
+
+        >>> arr.astype(np.dtype('float64'))
+        ... # doctest: +NORMALIZE_WHITESPACE
+        [0, 0, 1.0, 2.0]
+        Fill: 0
+        IntIndex
+        Indices: array([2, 3], dtype=int32)
+
+        Use a SparseDtype if you wish to be change the fill value as well.
+
+        >>> arr.astype(SparseDtype("float64", fill_value=np.nan))
+        ... # doctest: +NORMALIZE_WHITESPACE
+        [nan, nan, 1.0, 2.0]
+        Fill: nan
+        IntIndex
+        Indices: array([2, 3], dtype=int32)
         """
-        sp_values = self.sp_values
-        valid_spvals = np.isfinite(sp_values).sum()
-        if self._null_fill_value:
-            return valid_spvals
+        dtype = pandas_dtype(dtype)
+
+        if not isinstance(dtype, SparseDtype):
+            dtype = SparseDtype(dtype, fill_value=self.fill_value)
+
+        sp_values = astype_nansafe(self.sp_values,
+                                   dtype.subtype,
+                                   copy=copy)
+        if sp_values is self.sp_values and copy:
+            sp_values = sp_values.copy()
+
+        return self._simple_new(sp_values,
+                                self.sp_index,
+                                dtype)
+
+    def map(self, mapper):
+        # this is used in apply.
+        # We get hit since we're an "is_extension_type" but regular extension
+        # types are not hit...
```

Comment:
I can't seem to hit it anymore. I'll try removing and seeing if tests break.

Reply:
This is used in `Series.apply` for our internal extension types to avoid an `astype(object)`: https://github.com/pandas-dev/pandas/blob/c8ce3d01e9ffafc24c6f9dd568cd9eb7e42c610c/pandas/core/series.py#L3371-L3376

It's pretty useful for categorical and sparse.

Reply:
(on map) OK, that makes sense. It can then use a docstring though :)